Had a chance to test this on 0.108 and found there is a problem before the change

```
2020-04-06 19:27:22 ERROR (MainThread) [homeassistant.core] Error doing job: Future exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/homeassistant/homeassistant/components/tplink/light.py", line 231, in do_update_retry
    self._light_state = self.get_light_state_retry(self._light_features)
  File "/usr/src/homeassistant/homeassistant/components/tplink/light.py", line 294, in get_light_state_retry
    return self.get_light_state(light_features)
  File "/usr/src/homeassistant/homeassistant/components/tplink/light.py", line 310, in get_light_state
    brightness = brightness_from_percentage(self.smartbulb.brightness)
  File "/usr/local/lib/python3.7/site-packages/pyHS100/smartbulb.py", line 222, in brightness
    return int(light_state['brightness'])
KeyError: 'brightness'
2020-04-06 19:27:48 ERROR (MainThread) [homeassistant.core] Error doing job: Future exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/homeassistant/homeassistant/components/tplink/light.py", line 231, in do_update_retry
    self._light_state = self.get_light_state_retry(self._light_features)
  File "/usr/src/homeassistant/homeassistant/components/tplink/light.py", line 294, in get_light_state_retry
    return self.get_light_state(light_features)
  File "/usr/src/homeassistant/homeassistant/components/tplink/light.py", line 314, in get_light_state
    color_temp = kelvin_to_mired(self.smartbulb.color_temp)
  File "/usr/src/homeassistant/homeassistant/util/color.py", line 515, in color_temperature_kelvin_to_mired
    return math.floor(1000000 / kelvin_temperature)
ZeroDivisionError: division by zero
```

--- a/homeassistant/components/tplink/light.py
+++ b/homeassistant/components/tplink/light.py
@@ -86,10 +86,15 @@ class LightState(NamedTuple):
 
     def to_param(self):
         """Return a version that we can send to the bulb."""
+        if self.color_temp is not None:
+            color_temp = mired_to_kelvin(self.color_temp)
+        else:
+            color_temp = None
+
         return {
             "on_off": 1 if self.state else 0,
             "brightness": brightness_to_percentage(self.brightness),
-            "color_temp": mired_to_kelvin(self.color_temp) if self.color_temp else None,
+            "color_temp": color_temp,
             "hue": self.hs[0],
             "saturation": self.hs[1],
         }
@@ -151,18 +156,17 @@ class TPLinkSmartBulb(Light):
 
     async def async_turn_on(self, **kwargs):
         """Turn the light on."""
-        brightness = (
-            int(kwargs[ATTR_BRIGHTNESS])
-            if ATTR_BRIGHTNESS in kwargs
-            else self._light_state.brightness
-            if self._light_state.brightness is not None
-            else 255
-        )
-        color_tmp = (
-            int(kwargs[ATTR_COLOR_TEMP])
-            if ATTR_COLOR_TEMP in kwargs
-            else self._light_state.color_temp
-        )
+        if ATTR_BRIGHTNESS in kwargs:
+            brightness = int(kwargs[ATTR_BRIGHTNESS])
+        elif self._light_state.brightness is not None:
+            brightness = self._light_state.brightness
+        else:
+            brightness = 255
+
+        if ATTR_COLOR_TEMP in kwargs:
+            color_tmp = int(kwargs[ATTR_COLOR_TEMP])
+        else:
+            color_tmp = self._light_state.color_temp
 
         if ATTR_HS_COLOR in kwargs:
             # TP-Link requires integers.

--- a/homeassistant/components/tplink/light.py
+++ b/homeassistant/components/tplink/light.py
@@ -164,28 +164,26 @@ class TPLinkSmartBulb(Light):
             else self._light_state.color_temp
         )
 
+        if ATTR_HS_COLOR in kwargs:
+            # TP-Link requires integers.
+            hue_sat = tuple(int(val) for val in kwargs[ATTR_HS_COLOR])
+
+            # TP-Link cannot have both color temp and hue_sat
+            color_tmp = 0
+        else:
+            hue_sat = self._light_state.hs
+
         await self.async_set_light_state_retry(
             self._light_state,
-            LightState(
-                state=True,
-                brightness=brightness,
-                color_temp=color_tmp,
-                hs=tuple(kwargs.get(ATTR_HS_COLOR, self._light_state.hs or ())),
-                emeter_params=self._light_state.emeter_params,
+            self._light_state._replace(
+                state=True, brightness=brightness, color_temp=color_tmp, hs=hue_sat,
             ),
         )
 
     async def async_turn_off(self, **kwargs):
         """Turn the light off."""
         await self.async_set_light_state_retry(
-            self._light_state,
-            LightState(
-                state=False,
-                brightness=self._light_state.brightness,
-                color_temp=self._light_state.color_temp,
-                hs=self._light_state.hs,
-                emeter_params=self._light_state.emeter_params,
-            ),
+            self._light_state, self._light_state._replace(state=False),
         )
 
     @property

--- a/homeassistant/components/tplink/light.py
+++ b/homeassistant/components/tplink/light.py
@@ -89,7 +89,7 @@ class LightState(NamedTuple):
         return {
             "on_off": 1 if self.state else 0,
             "brightness": brightness_to_percentage(self.brightness),
-            "color_temp": mired_to_kelvin(self.color_temp),
+            "color_temp": mired_to_kelvin(self.color_temp) if self.color_temp else None,
             "hue": self.hs[0],
             "saturation": self.hs[1],
         }
@@ -393,7 +393,7 @@ class TPLinkSmartBulb(Light):
         diff = {
             key: value
             for key, value in new_state_param.items()
-            if new_state_param[key] != old_state_param[key]
+            if new_state_param.get(key) != old_state_param.get(key)
         }
 
         if not diff:

